File: javascript/hamming/hamming.js
// imperative solution
function compute(leftStrand = '', rightStrand = '') {
  if (leftStrand.length !== rightStrand.length) {
    throw new Error('left and right strands must be of equal length');
  }

  let diffCount = 0;
  if (leftStrand === rightStrand) return diffCount;

  for (let idx = 0; idx < leftStrand.length; idx += 1) {
    if (leftStrand.charAt(idx) !== rightStrand.charAt(idx)) {
      diffCount += 1;
    }
  }
  return diffCount;
}

// declarative solution
function computeDeclarative(leftStrand = '', rightStrand = '') {
  if (leftStrand.length === rightStrand.length) {
    if (leftStrand !== rightStrand) {
      return leftStrand.split('').filter((ch, idx) => ch !== rightStrand[idx])
        .length;
    }
    return 0;
  }
  throw new Error('left and right strands must be of equal length');
}

export { compute, computeDeclarative };

File: java/hamming/src/main/java/Hamming.java
import java.text.CharacterIterator;
import java.text.StringCharacterIterator;

public class Hamming {
  private int hammingDistance = 0;

  public Hamming(final String leftStrand, final String rightStrand) {
    this.validate(leftStrand, rightStrand);
    this.calculate(leftStrand, rightStrand);
  }

  public int getHammingDistance() {
    return this.hammingDistance;
  }

  private void validate(final String _leftStrand, final String _rightStrand) {
    String errorMessage = null;

    if (_leftStrand.length() != _rightStrand.length()) {
      errorMessage = "leftStrand and rightStrand must be of equal length.";

      if (_leftStrand.isEmpty())
        errorMessage = "left strand must not be empty.";
      if (_rightStrand.isEmpty())
        errorMessage = "right strand must not be empty.";
    }

    if (null != errorMessage)
      throw new IllegalArgumentException(errorMessage);
  }

  private void calculate(final String _leftStrand, final String _rightStrand) {
    CharacterIterator it = new StringCharacterIterator(_leftStrand);
    while (it.current() != CharacterIterator.DONE) {
      if (it.current() != _rightStrand.charAt(it.getIndex()))
        this.hammingDistance++;
      it.next();
    }
  }
}
